### PR TITLE
Improvement: 3DP profile category operation support

### PR DESCRIPTION
**Pull request by MaShizhen**

### CODE REVIEW

File: src/app/ui/views/DefinitionCreator.jsx [hunk 1]
```diff
@@ -1,175 +1,194 @@
...
+        case PRINTING_MANAGER_TYPE_MATERIAL:
+            // selectCategory = i18n._('key-Laser/ToolManger/ProfileCreator-Select material type');
+            // createItemDescribe = i18n._('key-Cnc/ToolManger/ProfileCreator-Create Carving Tool');
+            createCategoryDescribe = '创建材料类目_';
+            createCategorySubDescribe = '输入材料类目名称_';
+            createItemDescribe = '创建材料配置_';
+            createItemSubDescribe = '输入材料配置名称_';
+            selectCategory = '选择材料类目_';
+            categoryName = '新材料类目名称_';
+            itemName = '新材料配置名称_';
```
Comment (jane-rose) [inline src/app/ui/views/DefinitionCreator.jsx:52]: Use english

File: src/app/ui/views/ProfileManager/index.jsx [hunk 2]
```diff
@@ -48,67 +46,68 @@ function creatCateArray(optionList) {
...
+    const [definitionState, _setDefinitionState] = useState({
+        definitionForManager: allDefinitions.find(d => d.definitionId === activeDefinitionID),
+        selectedSettingDefaultValue: getDefaultDefinition(activeDefinitionID),
+        definitionOptions: [],
+        selectedName: '',
+        isCategorySelected: false,
+        renamingStatus: false,
+        cates: []
+    });
+    const setDefinitionState = (obj) => {
```
Comment (jane-rose) [inline src/app/ui/views/ProfileManager/index.jsx:59]: change this

File: src/app/ui/views/ProfileManager/index.jsx [hunk 3]
```diff
@@ -252,22 +238,22 @@ function ProfileManager({ optionConfigGroup, disableCategory = true, managerTitl
             const definitionForManager = definitionState?.definitionForManager;
             const isCategorySelected = definitionState?.isCategorySelected;
             let title = i18n._('key-Printing/ProfileManager-Create Profile');
-            let copyType = '', copyCategoryName = '', copyToolName = '';
+            let copyType = '', copyCategoryName = '', copyItemName = '';
 
             if (!isCreate) {
                 title = i18n._('key-Printing/ProfileManager-Copy Profile');
-                copyType = isCategorySelected ? 'Material' : 'Tool';
+                copyType = isCategorySelected ? 'Category' : 'Item';
```
Comment (jane-rose) [inline src/app/ui/views/ProfileManager/index.jsx:245]: `Tool -> Item`?

File: src/app/ui/widgets/CncLaserList/ToolPathList/ToolPathListBox.jsx [hunk 4]
```diff
@@ -170,11 +170,8 @@ const ToolPathListBox = (props) => {
                 check: !check
             }));
             if (displayedType === DISPLAYED_TYPE_TOOLPATH) {
-                let allToolPathHide = true;
-                toolPaths.forEach((toolPath) => {
-                    if ((toolPath.id === id && !visible) || (toolPath.id !== id && toolPath.visible)) {
-                        allToolPathHide = false;
-                    }
+                const allToolPathHide = toolPaths.some((toolPath) => {
+                    return !(toolPath.id === id && !visible) || (toolPath.id !== id && toolPath.visible);
```
Comment (jane-rose) [inline src/app/ui/widgets/CncLaserList/ToolPathList/ToolPathListBox.jsx:174]: check this

File: src/app/flux/editor/index.js [hunk 5]
```diff
@@ -1000,8 +1000,8 @@ export const actions = {
         const toolPaths = toolPathGroup.getToolPaths();
         const emptyToolPaths = [];
         toolPaths.forEach((item) => {
-            if (item.visibleModelIDs && item.visibleModelIDs.length) {
-                for (const id of item.visibleModelIDs) {
+            if (item.modelMap && item.modelMap.size) {
+                for (const [id] of item.modelMap) {
```
Comment (xiekun1992) [inline src/app/flux/editor/index.js:1004]: make it simpler to understand

File: src/app/ui/widgets/PrintingMaterial/MaterialWithColor.jsx [hunk 6]
```diff
@@ -4,7 +4,7 @@ import PropTypes from 'prop-types';
 export const MaterialWithColor = ({ name, color }) => {
     return (
         <div className="sm-flex align-center justify-space-between">
-            <span>{name}</span>
+            <span className="text-overflow-ellipsis" style={{ maxWidth: 'calc(100% - 16px)' }}>{name}</span>
```
Comment (xiekun1992) [inline src/app/ui/widgets/PrintingMaterial/MaterialWithColor.jsx:7]: define style in a css `class`

File: src/app/ui/widgets/CncLaserOutput/Output.jsx [hunk 7]
```diff
@@ -187,7 +176,7 @@ const Output = ({ headType }) => {
                         type="primary"
                         priority="level-one"
                         onClick={actions.preview}
-                        disabled={(!hasToolPathModel ?? false) || disablePreview}
+                        disabled={(!hasToolPathModel ?? false) || disableExport}
```
Comment (xiekun1992) [inline src/app/ui/widgets/CncLaserOutput/Output.jsx:179]: generate gcode and preview has anything to do with `disableExport`? maybe you need another variable?

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/flux/printing/index.js [hunk 8]
```diff
@@ -1040,17 +1042,72 @@ export const actions = {
...
+    removeToolCategoryDefinition: (type, category) => async (dispatch, getState) => {
+        const state = getState().printing;
+        const definitionsKey = defaultDefinitionKeys[type].definitions;
+
+        const definitions = state[definitionsKey];
+        const definitionsWithSameCategory = definitions.filter(d => d.category === category);
+        const ps = definitionsWithSameCategory.map((item) => {
+            return dispatch(actions.removeDefinitionByType(type, item));
+        });
+        return Promise.all(ps);
```
Comment (wen-qing-255) [inline src/app/flux/printing/index.js:1108]: `definitions` maybe can batch update

**Review by jane-rose — APPROVED** [review]

**Review by xiekun1992 — APPROVED** [review]

**Pull request merged**